### DISCUSSION TITLE
CSP directive blocking style src to load

### DISCUSSION THREAD

**Discussion by shubham-truboard** [mantinedev/mantine#6546] [Q&A]
Hi,

I'm using mantine version 6.0.13 and facing issue after applying the CSP directive in header, mentioned below

CSP 
**"Content-Security-Policy",
    "default-src http://localhost; object-src 'none'; img-src 'self' blob: http://localhost; require-trusted-types-for 'script'; script-src 'strict-dynamic' 'nonce-rAnd0m123' http://localhost; script-src-elem 'self' http://localhost; script-src-attr 'self' http://localhost:3000/; base-uri 'self' http://localhost:3000; style-src 'self' 'nonce-rAnd0m123' http://localhost:3000; font-src 'self';"**

Issue 

**Refused to apply inline style because it violates the following Content Security Policy directive: "style-src 'self' 'nonce-rAnd0m123' http://localhost:3000". Either the 'unsafe-inline' keyword, a hash ('sha256-H/s/dWGkGDaCkKqmo0VNeHrTgvJjinI5uvu7UmY6EB8='), or a nonce ('nonce-...') is required to enable inline execution. Note that hashes do not apply to event handlers, style attributes and javascript: navigations unless the 'unsafe-hashes' keyword is present.**

We have not used any inline styles  src in my code, but still getting the above mentioned error, might it looks like mantine creating internally a inline style src

Please provide some suggestion to fix this issue or any workaround.

**Comment by rtivital** [comment]
There might be some inline styles that are coming from radix-ui scrollarea. The only solution is to migrate to the latest 7.x version. As far as I know, there aren't any issues with CSP (multiple users reported to successfully implementing this on the Discord server).

**Reply by shubham-truboard** [reply]
Thanks for quick reply @rtivital, as suggested I updated the mantine to latest 7.11.2, now style src not getting blocked, but getting new error

**TypeError: Failed to set the 'innerHTML' property on 'Element': This document requires 'TrustedHTML' assignment.**

After rechecking fount that **require-trusted-types-for 'script';** blocking page to load
![Screenshot 2024-07-19 144801](https://github.com/user-attachments/assets/ec5720e8-a844-4204-ad35-9c06ef860c43)

Please suggest the solution for this

**Reply by rtivital** [reply]
I can't help with it. I've never seen such error before. Probably you CSP is to strict.

**Comment by uladzimirdev** [comment]
@rtivital I have a similar problem: v6 and CSP errors. I found out that the starting point in my case is https://github.com/mantinedev/mantine/blob/v6/src/mantine-core/src/Popover/PopoverDropdown/PopoverDropdown.tsx

<img width="493" alt="image" src="https://github.com/user-attachments/assets/26c1e370-5e62-4df1-a11c-3e935b9c7a8e">

any idea? I checked emotion code on the top of screenshot and it has access to `nonce` and the value is correct

**Reply by rtivital** [reply]
I cannot help with csp, I do not use it in my projects and I do not have enough knowledge on this topic.

**Reply by uladzimirdev** [reply]
@rtivital okay, a bit different question. I got in my app two types of mantine cache: emotion and mantine

and I use the next way of passing cache to Mantine

```
export const EmotionCacheProvider = ({
  children,
}: EmotionCacheProviderProps) => {
  const emotionCache = useMemo(() => {
    const insertionPoint = typeof document !== 'undefined'
      ? (document.querySelector('meta[name="emotion-insertion-point"]')! as HTMLElement)
      : undefined

    const cache = createCache({ key: "emotion", nonce: window.MetabaseNonce, insertionPoint });
    // This disables :first-child not working in SSR warnings
    // Source: https://github.com/emotion-js/emotion/issues/1105#issuecomment-557726922
    cache.compat = true;
    return cache;
  }, []);

  return <CacheProvider value={emotionCache}>{children}</CacheProvider>;
};

////////////////////

export const ThemeProvider = withEmotionCache(
  (props: ThemeProviderProps, cache: EmotionCache) => {
    // Merge default theme overrides with user-provided theme overrides
    const theme = useMemo(() => {
      return merge(getThemeOverrides(), props.theme);
    }, [props.theme]);

    return (
      <MantineProvider theme={theme} emotionCache={cache}>
        <DatesProvider>{props.children}</DatesProvider>
      </MantineProvider>
    );
  },
);
```

why can it happen?
<img width="458" alt="Screenshot 2024-08-08 at 00 30 41" src="https://github.com/user-attachments/assets/fa589ffc-76b7-468d-9337-7d9f1f7000c5">

<img width="528" alt="Screenshot 2024-08-08 at 00 30 36" src="https://github.com/user-attachments/assets/ed694a0d-ae4d-4391-86e4-c924ebb64fba">

**Reply by uladzimirdev** [reply]
I definitely hit this issue https://github.com/mantinedev/mantine/issues/3705

**Reply by rtivital** [reply]
You can use emotion CacheProvider component in addition to MantineProvider

**Reply by uladzimirdev** [reply]
I used both providers, but with the single cache instance. It looks like mantine has it's own cache created under the hood

**Reply by uladzimirdev** [reply]
it turned out this was a mistake on my side, we had some hidden rendering of the component to measure it's sizes and used Mantine provider without passing emotion cache there.

Thanks for your answers @rtivital!
